File: BACKEND/TaskSphere.API/Services/TaskService.cs
using Microsoft.EntityFrameworkCore;
using TaskSphere.API.DTOs;
using TaskSphere.API.Data;
using TaskSphere.API.Models;

namespace TaskSphere.API.Services
{
    public class TaskService
    {
        private readonly ApplicationDbContext _db;

        public TaskService(ApplicationDbContext db)
        {
            _db = db;
        }

        public async Task<List<TaskDto>> GetTasks(int userId)
        {
            return await _db.Tasks
                .Where(t => t.UserId == userId)
                .Select(t => new TaskDto
                {
                    Id = t.Id,
                    Title = t.Title,
                    Description = t.Description,
                    Status = t.Status,
                    Priority = t.Priority,
                    DueDate = t.DueDate
                })
                .ToListAsync();
        }

        public async Task<TaskDto?> GetTask(int id, int userId)
        {
            return await _db.Tasks
                .Where(t => t.Id == id && t.UserId == userId)
                .Select(t => new TaskDto
                {
                    Id = t.Id,
                    Title = t.Title,
                    Description = t.Description,
                    Status = t.Status,
                    Priority = t.Priority,
                    DueDate = t.DueDate
                })
                .FirstOrDefaultAsync();
        }

        public async Task<TaskDto> CreateTask(TaskDto dto, int userId)
        {
            var entity = new TaskItem
            {
                Title = dto.Title,
                Description = dto.Description,
                Status = dto.Status,
                Priority = dto.Priority,
                DueDate = dto.DueDate,
                UserId = userId
            };
            _db.Tasks.Add(entity);
            await _db.SaveChangesAsync();
            dto.Id = entity.Id;
            return dto;
        }

        public async Task<bool> UpdateTask(int id, TaskDto dto, int userId)
        {
            var entity = await _db.Tasks
                .FirstOrDefaultAsync(t => t.Id == id && t.UserId == userId);
            if (entity == null) return false;

            entity.Title = dto.Title;
            entity.Description = dto.Description;
            entity.Status = dto.Status;
            entity.Priority = dto.Priority;
            entity.DueDate = dto.DueDate;

            await _db.SaveChangesAsync();
            return true;
        }

        public async Task<bool> DeleteTask(int id, int userId)
        {
            var entity = await _db.Tasks
                .FirstOrDefaultAsync(t => t.Id == id && t.UserId == userId);
            if (entity == null) return false;

            _db.Tasks.Remove(entity);
            await _db.SaveChangesAsync();
            return true;
        }
    }
}

File: FRONTEND/tasksphere-frontend/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, catchError, throwError, timeout } from 'rxjs';
import { map } from 'rxjs/operators';

export interface Task {
  id: number;
  title: string;
  description: string;
  status: string;
  priority: string;
  dueDate?: string;
}

@Injectable({
  providedIn: 'root'
})
export class TaskService {

  public apiUrl = 'http://localhost:5148/api/Tasks';

  constructor(private http: HttpClient) {}

  getTasks(): Observable<Task[]> {
    return this.http.get<any>(this.apiUrl).pipe(
      map((res: any) => {
        if (Array.isArray(res)) {
          return res as Task[];
        }
        if (res && Array.isArray(res.tasks)) {
          return res.tasks as Task[];
        }
        return [] as Task[];
      }),
      catchError(err => {
        return throwError(() => err);
      })
    );
  }

  createTask(task: Partial<Task>): Observable<Task> {
    return this.http.post<Task>(this.apiUrl, task);
  }

  getTask(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.apiUrl}/${id}`).pipe(
      timeout(5000),
      catchError(err => {
        return throwError(() => err);
      })
    );
  }

  updateTask(id: number, changes: Partial<Task>): Observable<Task> {
    return this.http.put<Task>(`${this.apiUrl}/${id}`, changes);
  }

  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

}